### ISSUE TITLE
Name clash with HAP package - `StarGraph`

### ISSUE THREAD

**Issue by alex-konovalov** [digraphs/Digraphs#514]
First, congratulations with the new release 1.5.0 with the new databases.

It was picked up and it breaks GAP tests when all packages are loaded. It is not possible to use HAP package if Digraphs is loaded. This happens because of StarGraph which was added to Digraphs in https://github.com/digraphs/Digraphs/commit/de4a19772b9dad6c136f470a51e403c0929140bb and to HAP in 2016 in https://github.com/gap-packages/hap/commit/0e03ed2cab330d97b02750d0aea9fa3d2c59af71.

What happens is
```
#I  HAP: start reading file 'init.g'
Error, operation `StarGraph' has been declared as a one argument Operation and cannot also be an Attribute  
```

Perhaps @grahamknockillaree could help to resolve this? It seems undocumented and only used in HAP examples.

**Comment by wilfwilson** [comment]
Thanks for mentioning this @alex-konovalov

**Comment by alex-konovalov** [comment]
@wilfwilson We need to resolve this to be able to include new Digraphs in the next GAP distribution.

**Comment by grahamknockillaree** [comment]
I'll look into this tonight. (I have a full schedule today.)

Graham


School of Mathematics, Statistics & Applied Mathematics
National University of Ireland, Galway
tel: 091 493011



________________________________

**Comment by grahamknockillaree** [comment]
In HAP:  StarGraph(G) is an operation which inputs a finitely presented group G and returns the star graph of its presentation.

In digraphs: StarGraph( [filt, ]k ) is an operation which inputs an integer k, and possible an option, and returns a graph (according to the online manual for Digraphs 1.5.0) .

I'm not sure why this should cause an error, or why the error mentions "Attribute".

**Comment by wilfwilson** [comment]
HAP declares `StarGraph` as an attribute on `lib/hap.gd`, line 362:
https://github.com/gap-packages/hap/blob/8881bab2ed3f2ad312d2e519656e664a929f5f69/lib/hap.gd#L362
```
 DeclareAttribute("StarGraph",IsFpGroup); 
```
This is why the error occurs: it is not possible in GAP for a name to be both an operation and an attribute.

I presume that `StarGraph` was declared as such in order to retain the computed value of the star graph for the fp-group group, it seems reasonable to me.

**Comment by grahamknockillaree** [comment]
Now I understand -- I think HAP probably shouldn't use a name like StarGraph as an attribute.  Let me figure out how easy it is to make StarGraph just an operation.  I'll take a look at this tonight.

**Comment by wilfwilson** [comment]
One option would be to do something like the following. It would retain the storing-the-value-of-`StarGraph` functionality while keeping `StarGraph` as an operation:

Declare a 'hidden' attribute and the operation:
```
DeclareAttribute("StarGraphAttr", IsFpGroup);
DeclareOperation("StarGraph", [IsFpGroup]);
```
Then install methods like the following:
```
# This means that StarGraphAttr delegates to StarGraph
InstallMethod(StarGraphAttr, [IsFpGroup], StarGraph);

# The StarGraph method exists as before, except that it checks/sets the value of the attribute as appropriate
InstallMethod(StarGraph, [IsFpGroup],
function(G)
  if HasStarGraphAttr(G) then
    return StarGraphAttr(G);
  fi;
  # insert existing code to compute StarGraph here, except for the return statement
  SetStarGraphAttr(G, graph);
  return graph;
end);
```

**Comment by wilfwilson** [comment]
This is now resolved in HAP 1.38.
